### ISSUE TITLE
The `rollbackCurrentTransaction` method does not rollback transactions by the time it returns

### ISSUE THREAD

**Issue by liu-ronny** [chax-at/transactional-prisma-testing#15]
## Issue

The `rollbackCurrentTransaction` method is used to end an active transaction. However, it is possible for the rollback to not be complete by the time the function returns, leading to a scenario in which a `ROLLBACK` command is not issued before additional code runs.

## Example

The below code demonstrates the issue. 

```ts
import { PrismaTestingHelper } from '@chax-at/transactional-prisma-testing';
import { PrismaClient } from '@prisma/client';

let prisma = new PrismaClient({
  log: ['query'],
});

const prismaTestingHelper = new PrismaTestingHelper(prisma);

const beginTx = async () => {
  await prisma.$connect();
  await prismaTestingHelper.startNewTransaction();
};

const rollbackTx = async () => {
  prismaTestingHelper.rollbackCurrentTransaction();
  await prisma.$disconnect();
};

prisma = prismaTestingHelper.getProxyClient();

await beginTx();
await prisma.$executeRaw`select 1`;
await rollbackTx();
```

With query logs turned on, I consistently get the following output, which is missing a `ROLLBACK` command:

```
prisma:query BEGIN
prisma:query select 1
```

I can also reproduce the issue with the below code. Here, a query that shouldn't be part of the transaction manages to sneak in. However, it's difficult to consistently trigger. It seems to only work once out of many tries.

```ts
import { PrismaTestingHelper } from '@chax-at/transactional-prisma-testing';
import { PrismaClient } from '@prisma/client';

let prisma = new PrismaClient({
  log: ['query'],
});

const prismaTestingHelper = new PrismaTestingHelper(prisma);
prisma = prismaTestingHelper.getProxyClient();

await prismaTestingHelper.startNewTransaction();
await prisma.$executeRaw`select 1`;
prismaTestingHelper.rollbackCurrentTransaction();

// This should not run within the previous transaction.
await prisma.$executeRaw`select 1`;
```

Query logs:

```
prisma:query BEGIN
prisma:query select 1
prisma:query select 1
prisma:query ROLLBACK
```

## Analysis

When `rollbackCurrentTransaction` is called, it resolves the inner promise shown below. This action allows the `$transaction` callback to continue executing, ultimately throwing an error that rolls back the transaction. However, before the `$transaction` callback can throw its error, the rest of`rollbackCurrentTransaction` runs, and statements that come after may be run before the transaction actually ends.

https://github.com/chax-at/transactional-prisma-testing/blob/441b51d89d3498ce383b548d6f7a975b2f59cda0/src/prisma.testing.helper.ts#L202-L211

## Potential Solution

`rollbackCurrentTransaction` probably needs to be async to work as intended. Here's one way it could be changed. I believe existing code that runs `rollbackCurrentTransaction` synchronously should still work the same. But the method would need to be awaited to avoid subtle bugs.

```ts
  /**
   * Starts a new transaction and automatically updates the proxy client (no need to fetch it again).
   * Must be called before each test.
   */
  public async startNewTransaction(opts?: { timeout?: number; maxWait?: number }): Promise<void> {
    if(this.endCurrentTransactionPromise != null) {
      throw new Error('[transactional-prisma-testing] rollbackCurrentTransaction must be called before starting a new transaction');
    }
    this.savepointId = 0;
    // This is a workaround for https://github.com/prisma/prisma/issues/12458
    return new Promise(resolve => {
      this.prismaClient.$transaction(async prisma => {
        this.currentPrismaTransactionClient = prisma as Prisma.TransactionClient | undefined;
        await new Promise((innerResolve, innerReject) => {
          this.endCurrentTransactionPromise = () => {
            // We intentionally want to do a rollback of the transaction after a successful run
            innerReject(internalRollbackErrorSymbol);
          };
          resolve();
        });
      }, opts).catch((error) => {
        if(error !== internalRollbackErrorSymbol || !this.endRollbackIssuedPromise) {
          throw error;
        }

        // Once we catch an error, we know Prisma has rolled back the transaction
        this.endRollbackIssuedPromise();
        this.endRollbackIssuedPromise = undefined;
      });
    });
  }

  /**
   * Ends the currently active transaction. Must be called after each test.
   */
  public async rollbackCurrentTransaction(): void {
    if(this.endCurrentTransactionPromise == null) {
      throw new Error('[transactional-prisma-testing] No transaction currently active');
    }

    // Wait for the `ROLLBACK` command to be issued before proceeding
    const rollbackIssued = new Promise((resolve) => {
      this.endRollbackIssuedPromise = resolve;
    });

    this.endCurrentTransactionPromise();
    this.endCurrentTransactionPromise = undefined;

    await rollbackIssued;
  }
```

**Comment by Valerionn** [comment]
Hey!

First of all, thanks for the report, it's always nice to receive such a well-written issue report.

However, the behavior you're describing is mostly "by design". While it seems weird, it should not lead to actual problems in real applications (but if you have a real-life example for this being a problem, then I can of course change it).

Regarding your examples:

In your first example, you're immediately calling `prisma.$disconnect()`. This can indeed lead to the `ROLLBACK` command not being sent before the connection is closed - however, Postgres will automatically rollback your transaction when the client disconnects (since the transaction has not been commited), leading to the same result.

In your second example, you're using the package in a non-intended way by sending a request after calling `rollbackCurrentTransaction` and before calling `startNewTransaction`. In this case, the call is _always_ redirected to the transaction client, probably leading to an error if it was rollbacked already. The intended way would be this:
```ts
const prismaTestingHelper = new PrismaTestingHelper(prisma);
prisma = prismaTestingHelper.getProxyClient();

await prismaTestingHelper.startNewTransaction();
await prisma.$executeRaw`select 1`;
prismaTestingHelper.rollbackCurrentTransaction();

// v--------  This call is missing
await prismaTestingHelper.startNewTransaction();

// <--------- This will now run in the new transaction, even if the old one is not rollbacked yet
await prisma.$executeRaw`select 1`;
```
There is still a potential race condition in which the second `SELECT 1` query will be sent before the transaction is rollbacked - but then, the second query will be executed inside the second transaction while the first transaction will be rollbacked in the background. But this should be totally fine since those two transactions are independent (and shouldn't be a long overlap anyway).

I do agree though that this behavior is a bit weird - but I don't know what the "correct" behavior in this case would be. Probably redirecting the call to the original prisma client? Though this can lead to subtle unintended behavior as well, I think, though it's possible to do this, even without having to await the rollback call.

The two main reasons why the rollback function doesn't return a Promise are:
1. Performance: It's faster to not wait for the rollback since it can happen in the background
2. When implementing it, I actually thought about doing it the way you suggested - but I don't know how Prisma handles this exactly and if the `.catch(` will actually only be called _after_ the rollback. Since it didn't lead to any observable problems for me, I never bothered testing Prisma's internal behavior in this case

Again, if you're actually observing any issues with this behavior in a real-world application, then please describe them, otherwise I would tend to keep the current behavior (and only think about maybe redirecting calls between `rollback` and `startNew` to the original client instead of the previous transaction client)

**Comment by liu-ronny** [comment]
Hey, thanks for the kind words!

Your points are fair - I haven't observed any meaningful issues in my application tests. I only stumbled upon the behavior while debugging an unrelated problem and realized I wasn't seeing the `ROLLBACK` command in the query logs. Figured it was worth reporting just in case.

Regarding your reason #2:

When an error is thrown within an interactive transaction callback, Prisma seems to do the following:

1. Catch the error
2. Silently issue a `ROLLBACK`
3. Re-throw the original error

What we catch is the re-thrown error, and by the time we're in our own catch callback, the `ROLLBACK` should be issued.

Source:

https://github.com/prisma/prisma/blob/1eb333d55da0133679a4f63de47aba6bda823496/packages/client/src/runtime/getPrismaClient.ts#L783-L816

In any case, since I haven't observed any real-world problems with the current behavior, I don't feel particularly compelled to push for any changes. Do you think it's best to just close the issue?

**Comment by Valerionn** [comment]
@liu-ronny Thanks for the investigation, that's interesting to know! Yeah, I'll close the issue for now - but if you encounter any issues with this behavior, then please open a new issue and then I can adapt it according to your findings here.
